fdbserver/ptxn/TLogServer.actor.cpp

```diff
@@ -954,6 +959,66 @@ ACTOR Future<Void> tLogCommit(Reference<TLogGroupData> self,
 	return Void();
 }
 
+ACTOR Future<Void> tLogPeekMessages(Reference<TLogGroupData> self,
+                                    TLogPeekRequest req,
+                                    Reference<LogGenerationData> logData) {
+	state TLogPeekReply reply;
+	state TLogStorageServerMessageSerializer serializer(req.storageTeamID);
+	state Reference<LogGenerationData::StorageTeamData> storageTeamData = logData->getStorageTeamData(req.storageTeamID);
+
+	if (storageTeamData) {
+		if (storageTeamData->versionMessages.empty() ||
+		    std::get<Version>(storageTeamData->versionMessages.back()) < req.beginVersion) {
+			wait(logData->version.whenAtLeast(req.beginVersion));
+		}
+
+		for (const auto& tuple : storageTeamData->versionMessages) {
+			const Version& version = std::get<Version>(tuple);
+			if (version < req.beginVersion) {
+				continue;
+			}
+			if (req.endVersion.present() && version > req.endVersion.get()) {
+				break;
+			}
+			if (!reply.begin.present()) {
+				reply.begin = version;
+			}
+			reply.end = version;
+
+			serializer.startVersionWriting(version);
+			const StringRef& messages = std::get<StringRef>(tuple);
+			const Arena& arena = std::get<Arena>(tuple);
+
+			// TODO: instead of deserializing, just sends it out
+			ProxyTLogMessageHeader header;
+			std::vector<SubsequenceMutationItem> items;
+			// TODO: deserialization can throw incompatible_protocol_version() error.
+			bool b = proxyTLogPushMessageDeserializer(arena, messages, header, items);
+			if (!b) {
+				std::cout << "Error deserializing for version " << version << " \n";
+				continue;
+			}
+			for (const auto& item : items) {
+				serializer.writeSubsequenceMutationRef(item);
+			}
+			serializer.completeVersionWriting();
+		}
+	} else {
+		// TODO: block here or return error?
+	}
+	serializer.completeMessageWriting();
+	Standalone<StringRef> buffer = serializer.getSerialized();
+
+	reply.arena = buffer.arena();
+	reply.data = buffer;
+	reply.maxKnownVersion = logData->version.get();
+	reply.minKnownCommittedVersion = logData->minKnownCommittedVersion;
+	// reply.onlySpilled = false;
```

Comment:
feel free to remove variable that doesn't make sense now, like 'onlySpilled'.

Reply:
Sure. I left here in case we'll use the field in the future.